airflow/migrations/versions/c306b5b5ae4a_switch_xcom_table_to_use_run_id.py

```diff
@@ -0,0 +1,140 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Switch XCom table to use ``run_id``.
+
+Revision ID: c306b5b5ae4a
+Revises: e655c0453f75
+Create Date: 2022-01-19 03:20:35.329037
+"""
+from typing import Sequence
+
+from alembic import op
+from sqlalchemy import Column, Integer, LargeBinary, MetaData, Table, select
+
+from airflow.migrations.db_types import TIMESTAMP, StringID
+
+# Revision identifiers, used by Alembic.
+revision = "c306b5b5ae4a"
+down_revision = "e655c0453f75"
+branch_labels = None
+depends_on = None
+
+
+metadata = MetaData()
+
+dagrun = Table(
+    "dag_run",
+    metadata,
+    Column("id", Integer, primary_key=True),
+    Column("dag_id", StringID(), nullable=False),
+    Column("run_id", StringID(), nullable=False),
+    Column("execution_date", TIMESTAMP, nullable=False),
+)
+
+
+def _get_new_xcom_columns() -> Sequence[Column]:
+    return [
+        Column("dagrun_id", Integer(), nullable=False),
+        Column("task_id", StringID(), nullable=False),
+        Column("key", StringID(length=512), nullable=False),
+        Column("value", LargeBinary),
+        Column("timestamp", TIMESTAMP, nullable=False),
+        Column("dag_id", StringID(), nullable=False),
+        Column("run_id", StringID(), nullable=False),
+    ]
+
+
+def _get_old_xcom_columns() -> Sequence[Column]:
+    return [
+        Column("key", StringID(length=512), nullable=False),
+        Column("value", LargeBinary),
+        Column("timestamp", TIMESTAMP, nullable=False),
+        Column("task_id", StringID(), nullable=False),
+        Column("dag_id", StringID(), nullable=False),
+        Column("execution_date", StringID(), nullable=False),
+    ]
+
+
+def upgrade():
+    """Switch XCom table to use run_id.
+
+    For performance reasons, this is done by creating a new table with needed
+    data pre-populated, adding back constraints we need, and renaming it to
+    replace the existing XCom table.
+    """
```

Comment:
I don't think this will work with offline generation of sql

Reply:
Hmm, why not? I was assuming it would because `op` ultimately receives a raw SQL string and can “generate” it verbatim.

Reply:
I'm worried about the `execute`. Won't this work with the format used here: https://github.com/apache/airflow/blob/8505d2f0a4524313e3eff7a4f16b9a9439c7a79f/airflow/migrations/versions/2e42bb497a22_rename_last_scheduler_run_column.py#L42-L45

Reply:
Yeah this is fine -- as everything is going via `op` and `batch_op` -- the main thing is nothing accessing a Session directly.